src/osd/OSD.h

```diff
@@ -1631,8 +1631,7 @@ class OSD : public Dispatcher,
   map<int, int> debug_heartbeat_drops_remaining;
   Cond heartbeat_cond;
   bool heartbeat_stop;
-  Mutex heartbeat_update_lock; // orders under heartbeat_lock
-  bool heartbeat_need_update;   ///< true if we need to refresh our heartbeat peers
+  std::atomic_bool heartbeat_need_update;   
```

Comment:
i'd recommend use `std::atomic_flag` instead, `std::atomic<>` would use lock under the hood. but `std::atomic_flag` is lock free.